### DISCUSSION TITLE
Incompatible packages advice

### DISCUSSION THREAD

**Discussion by hughetop** [AcademySoftwareFoundation/rez#1375] [Q&A]
Hello,

I'm running into another issue with maya (big surprise) in that it's special PySide2 package is incompatible with the normal industry-standard PySide2 package. What I mean by that is: if you have a custom PySide2 tool and want to launch it _within maya_ it must use maya's PySide2 package, otherwise it will crash. And conversely, if you try to launch it from a terminal, it _must_ use the normal PySide2 package, _not_ maya's PySide2.

Is there a good way to control the PySide2 import at runtime or is the recommendation to create 2 near-identical environments: one which requires maya, and another which requires PySide2?

platform-linux
os-centos-7.9.2009
arch-x86_64
maya-2022
PySide2-5.15.2.1

**Comment by hughetop** [comment]
Forgot to mention, another idea we had was to only require maya, but when launching from a terminal, patch the environment to include PySide2. I haven't played around with the rez python package, is there a clean way to do this with the rez python package? Something like a context manager ideally so that it doesn't modify the environment.

**Comment by maxnbk** [comment]
Ultimately, the "provides" feature is the solution to this. (discussed heavily elsewhere, but unimplemented)

Lacking that feature, the way I've seen this solved this in the past was to make the pyside-2 rez package have variants for maya, which, when resolved, point to mayas pyside stuff.

The downside is that sometimes you might want to ask for a vanilla pyside package, but receive a maya solve, which is subpar, and have to provide negations of `!maya` in requests in order to get the correct one. This has lead to a variety of unique situations, most of which are solvable, but it's definitely a subpar situation, and contributes to motivations to solve the unimplemented 'provides' feature.

**Reply by hughetop** [reply]
Interesting, I hadn't thought of using maya as a variant of PySide2. The trouble is that we would ideally want an environment with _both_ maya and PySide2, but choose which one to import based on whether it's launched from maya or a terminal. To make things more complicated, we're using Qt.py, so the actual PySide2 import is abstracted to allow for PyQt5 compatibility. We might just end up creating 2 environments, one with maya, and one with PySide2 in the short term.

**Reply by JeanChristopheMorinPerso** [reply]
@hughetop If you have control over your PySide2 package, you can always do something like:
```python
def commands():
    if 'maya' not in request:
        env.PATH.append('<path to PySide>')
```
It's dirty but it works.

**Reply by hughetop** [reply]
We tried something similar, by _prepending_ maya's PySide2 lib rather than _append_, so that it will always be found in the $PYTHONPATH first. That works fine when launching from within maya, but it crashes when launching from a terminal. At the moment, we basically can't reliably launch it from within maya and from a terminal in the same rez environment.

**Reply by maxnbk** [reply]
Are you using the correct python interpreter in each case?
If the maya pyside is prepended, you would only be able to use it when using mayapy rather than your vanilla python terminal, in that environment.

**Reply by hughetop** [reply]
OH! I will need to double check, but I'm pretty sure that's it. We were using mayapy in terminals for some reason, but switching to python should fix that.

**Reply by hughetop** [reply]
Well...you're mostly right, but apparently if vanilla PySide2 is in the environment _at all_, maya will crash. So PySide2 cannot be in the $PYTHONPATH if maya is launched. This seems to be an issue with maya, not rez or anything else. We've decided for now to just make 2 software environments, one with maya and another for the terminal.
